### PR TITLE
Upgrade: cosmos-sdk@v0.47, wasmd@v0.46, ibc-go@v7,

### PR DESCRIPTION
**Pull request by hard-nett**
currently, Bitsong does not support cosmwasm contracts with `cosmwasm1_4` capabilities. To add support, we needed atleast v0.42 of the wasmd module dependency, however with anticipation for refactoring Bitsong to support the latest versions of ibc & cosmos-sdk, this also updates the sdk to the latest version of v47, prior to v.50. This includes the packet-forward-middleware module

most damage and mess was made in cmd config & testing during refactor :)

update: 10/8/24:
- any new keypairs generated with this new version will not be compatible with older versions:
`err for key deploy-key.info: "Bytes left over in UnmarshalBinaryLengthPrefixed, should read 10 more bytes but have 157"`
- nodes MUST have minimum gas price set in `app.toml` prior to upgrade, or else node panics on upgrade
- nodes MUST use a Go version >= `v1.22.4`

### CODE REVIEW

**Comment by angelorc** [comment]
We can disable the merkledrop module. Instead of using the module we could use the cw-merkledrop smart contract

File: .gitignore [hunk 1]
```diff
@@ -28,6 +24,9 @@ mytestnet
 coverage.txt
 profile.out
 
+# heighliner
+heighliner
```
Comment (freak12techno) [inline .gitignore:28]: this line is stated twice, also on line 53

File: app/keepers/keepers.go [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+		govModAddress,
+	)
+
+	// Create Transfer Stack
+	var transferStack porttypes.IBCModule
+	transferStack = transfer.NewIBCModule(appKeepers.TransferKeeper)
+	transferStack = packetforward.NewIBCMiddleware(
+		transferStack,
+		appKeepers.PacketForwardKeeper,
+		0, // retries on timeout
```
Comment (freak12techno) [inline app/keepers/keepers.go:302]: can we move this to constant or something? this might seem highly unobvious

File: app/modules.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+func simulationModules(
+	app *BitsongApp,
+	encodingConfig encparams.EncodingConfig,
+	_ bool,
+) []module.AppModuleSimulation {
+	appCodec := encodingConfig.Marshaler
+
+	// bondDenom := app.GetChainBondDenom()
```
Comment (freak12techno) [inline app/modules.go:168]: maybe remove this?

File: app/upgrades/v018/upgrades.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+			case fantokentypes.ModuleName:
+				keyTable = fantokentypes.ParamKeyTable() //nolint:staticcheck
+				if !subspace.HasKeyTable() {
+					subspace.WithKeyTable(keyTable)
+				}
+			// wasmd module
+			case wasmtypes.ModuleName:
+				keyTable = wasmtypes.ParamKeyTable() //nolint:staticcheck
+			default:
+				continue
```
Comment (freak12techno) [inline app/upgrades/v018/upgrades.go:105]: do we have other modules to migrate? maybe add `logger.Info()` if we do not migrate something?

File: cmd/bitsongd/cmd/config.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			return errUnknownConfigKey(key)
+		}
+
+		confFile := filepath.Join(configPath, "client.toml")
+		if err := writeConfigToFile(confFile, &tcc); err != nil {
+			return fmt.Errorf("could not write client config to the file: %v", err)
+		}
+
+	default:
+		panic("cound not execute config command")
```
Comment (freak12techno) [inline cmd/bitsongd/cmd/config.go:154]: can we update it to show how to use the command instead?

File: cmd/bitsongd/cmd/genesis.go [hunk 6]
```diff
@@ -93,16 +90,16 @@ func MainnetGenesisParams() GenesisParams {
...
-	genParams.GovParams.DepositParams.MinDeposit = sdk.NewCoins(sdk.NewCoin(
-		appparams.MicroCoinUnit,
-		sdk.NewInt(512_000_000),
-	))
-	genParams.GovParams.TallyParams.Quorum = sdk.MustNewDecFromStr("0.4")          // 40%
-	genParams.GovParams.TallyParams.Threshold = sdk.MustNewDecFromStr("0.5")       // 50%
-	genParams.GovParams.TallyParams.VetoThreshold = sdk.MustNewDecFromStr("0.334") // 33.40%
-	genParams.GovParams.VotingParams.VotingPeriod = time.Hour * 24 * 7             // 7 days
+	genParams.GovParams = govtypes.DefaultConfig()
+	// genParams.GovParams.DepositParams.MaxDepositPeriod = time.Hour * 24 * 15 // 15 days
```
Comment (freak12techno) [inline cmd/bitsongd/cmd/genesis.go:94]: do we need this?

File: cmd/bitsongd/cmd/genesis.go [hunk 7]
```diff
@@ -130,90 +125,81 @@ func TestnetGenesisParams() GenesisParams {
 
 	genParams.StakingParams.UnbondingTime = time.Hour * 24 * 7 * 2 // 2 weeks
 
-	genParams.GovParams.DepositParams.MinDeposit = sdk.NewCoins(sdk.NewCoin(
-		appparams.MicroCoinUnit,
-		sdk.NewInt(1000000), // 1 BTSG
-	))
-	genParams.GovParams.TallyParams.Quorum = sdk.MustNewDecFromStr("0.0000000001") // 0.00000001%
-	genParams.GovParams.VotingParams.VotingPeriod = time.Second * 300              // 300 seconds
+	// genParams.GovParams.DepositParams.MinDeposit = sdk.NewCoins(sdk.NewCoin(
```
Comment (freak12techno) [inline cmd/bitsongd/cmd/genesis.go:128]: same, do we need this?

File: cmd/bitsongd/cmd/genesis.go [hunk 8]
```diff
@@ -130,90 +125,81 @@ func TestnetGenesisParams() GenesisParams {
...
-	govGenState := govtypes.DefaultGenesisState()
-	govGenState.DepositParams = genesisParams.GovParams.DepositParams
-	govGenState.TallyParams = genesisParams.GovParams.TallyParams
-	govGenState.VotingParams = genesisParams.GovParams.VotingParams
-	govGenStateBz, err := cdc.MarshalJSON(govGenState)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to marshal gov genesis state: %w", err)
-	}
-	appState[govtypes.ModuleName] = govGenStateBz
+	// govGenState := govtypes.DefaultGenesisState()
```
Comment (freak12techno) [inline cmd/bitsongd/cmd/genesis.go:169]: same

File: cmd/bitsongd/cmd/genesis.go [hunk 9]
```diff
@@ -261,9 +246,12 @@ Example:
 
 			// run Prepare Genesis
 			appState, genDoc, err = PrepareGenesis(clientCtx, appState, genDoc, genesisParams, chainID)
+			if err != nil {
+				return fmt.Errorf("err, %s", err.Error())
```
Comment (freak12techno) [inline cmd/bitsongd/cmd/genesis.go:250]: why not return err here?

File: go.mod [hunk 10]
```diff
@@ -1,159 +1,196 @@
...
 
 replace (
 	// dragonberry
 	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
-	// Comet
-	github.com/tendermint/tendermint => github.com/cometbft/cometbft v0.34.29
+// Comet
+
+// google.golang.org/grpc => google.golang.org/grpc v1.33.2
```
Comment (freak12techno) [inline go.mod:194]: maybe remove this as well?

File: app/upgrades/v018/upgrades.go (hunk 4 above)
Comment (hard-nett) [reply app/upgrades/v018/upgrades.go:105]: No. The only custom module bitsong has after this migration is the fantoken. The minimum sdk modules migrated were referenced here: 
- https://github.com/cosmos/cosmos-sdk/blob/0f99829fabe34ae489d71a8524fb4c82632965e3/simapp/upgrades.go 

here are some additional references for modules needed migation:
- osmosis: https://github.com/osmosis-labs/osmosis/blob/main/app/upgrades/v21/upgrades.go#L73
- juno: https://github.com/CosmosContracts/juno/blob/main/app/upgrades/v16/upgrades.go#L55

File: cmd/bitsongd/cmd/genesis.go (hunk 6 above)
Comment (hard-nett) [reply cmd/bitsongd/cmd/genesis.go:94]: not required, but needed if we want bitsong-go app to always set these values when creating new app.

re-implemented in upcoming iteration

File: cmd/bitsongd/cmd/genesis.go (hunk 7 above)
Comment (hard-nett) [reply cmd/bitsongd/cmd/genesis.go:128]: same as previous comment

File: CHANGELOG.md [hunk 11]
```diff
@@ -34,6 +34,31 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 # Changelog
 
+## [v0.17.0]
```
Comment (angelorc) [inline CHANGELOG.md:37]: This should be 0.18.x...

File: README.md [hunk 12]
```diff
@@ -16,15 +16,17 @@ After launching the main BitSong blockchain in August 2020, the bitsong-2b mainn
...
 There are many ways you can install BitSong Blockchain Testnet node on your machine.
 
 ## From Source
-1. **Install Go** 
+1. **Install Go & Environment Tooling** 
     ```bash
+	sudo apt update
+	sudo apt install -y curl git jq lz4 build-essential
     wget -q -O - https://git.io/vQhTU | bash -s -- --remove
     wget -q -O - https://git.io/vQhTU | bash -s -- --version 1.19.5
```
Comment (angelorc) [inline README.md:31]: replace with 1.22.9

File: app/app.go [hunk 13]
```diff
@@ -2,254 +2,185 @@ package app
...
-	merkledroptypes "github.com/bitsongofficial/go-bitsong/x/merkledrop/types"
+
+	"github.com/bitsongofficial/go-bitsong/v018/app/keepers"
+	"github.com/bitsongofficial/go-bitsong/v018/app/upgrades"
+	v010 "github.com/bitsongofficial/go-bitsong/v018/app/upgrades/v010"
+	v011 "github.com/bitsongofficial/go-bitsong/v018/app/upgrades/v011"
+	v013 "github.com/bitsongofficial/go-bitsong/v018/app/upgrades/v013"
+	v014 "github.com/bitsongofficial/go-bitsong/v018/app/upgrades/v014"
+	v015 "github.com/bitsongofficial/go-bitsong/v018/app/upgrades/v015"
+	v016 "github.com/bitsongofficial/go-bitsong/v018/app/upgrades/v016"
```
Comment (angelorc) [inline app/app.go:13]: I think we should remove all older versions

File: cmd/bitsongd/cmd/config.go (hunk 5 above)
Comment (hard-nett) [reply cmd/bitsongd/cmd/config.go:154]: this default is a placeholder, and really wont be accessed.

File: cmd/bitsongd/cmd/genesis.go (hunk 9 above)
Comment (hard-nett) [reply cmd/bitsongd/cmd/genesis.go:250]: the error is returned

File: app/app.go (hunk 13 above)
Comment (hard-nett) [reply app/app.go:13]: afaik, these are necessary for for the upgrade store keys, so we would run into unexpected issues if these are removed.

**Pull request merged**